test/core/xds/xds_transport_fake.cc

```diff
@@ -120,15 +124,45 @@ void FakeXdsTransportFactory::FakeStreamingCall::CompleteSendMessageFromClient(
   CompleteSendMessageFromClientLocked(ok);
 }
 
+void FakeXdsTransportFactory::FakeStreamingCall::StartRecvMessage() {
+  absl::optional<std::string> pending;
+  ReleasableMutexLock lock(&mu_);
+  ASSERT_FALSE(read_pending_);
+  read_count_ += 1;
+  read_pending_ = true;
+  if (!to_client_messages_.empty()) {
+    // Dispatch pending message (if there's one) on a separate thread to avoid
+    // recursion
+    std::thread thread(
+        [](const RefCountedPtr<FakeStreamingCall>& call) {
+          call->DispatchPendingMessageToClient();
+        },
+        Ref());
+    thread.detach();
+  }
+}
+
 void FakeXdsTransportFactory::FakeStreamingCall::SendMessageToClient(
     absl::string_view payload) {
-  ExecCtx exec_ctx;
-  RefCountedPtr<RefCountedEventHandler> event_handler;
-  {
-    MutexLock lock(&mu_);
-    event_handler = event_handler_->Ref();
+  ReleasableMutexLock lock(&mu_);
+  to_client_messages_.emplace_back(payload);
+  lock.Release();
+  DispatchPendingMessageToClient();
+}
+
+void FakeXdsTransportFactory::FakeStreamingCall::
+    DispatchPendingMessageToClient() {
+  ReleasableMutexLock lock(&mu_);
```

Comment:
Similar to above, please use an inner scope and `MutexLock` instead of `ReleaseableMutexLock`.